tensorflow/compiler/jit/xla_compilation_cache.cc

```diff
@@ -335,12 +346,128 @@ void LogOnceXlaCompiledFirstCluster() {
+}
+
+Status XlaCompilationCache::CompileAsynchronous(
+    Entry* entry, const XlaCompiler::Options& options,
+    const std::vector<XlaCompiler::Argument>& args,
+    const string &function_name,
+    const std::function<Status(XlaCompiler* compiler,
+                               const std::vector<XlaCompiler::Argument>& args,
+                               XlaCompiler::CompilationResult*)>& compile_fn) {
+  // Explicitly capture all required data by value for async compilation.
+  struct AsyncCompileData {
+    Entry* entry;
+    const XlaCompiler::Options& options;
+    const std::vector<XlaCompiler::Argument>& args;
+    const string &function_name;
+    const std::function<Status(XlaCompiler* compiler,
+                               const std::vector<XlaCompiler::Argument>& args,
+                               XlaCompiler::CompilationResult*)>& compile_fn;
+  } async_compile_data = {
+    entry, options, args, function_name, compile_fn
+  };
+
+  entry->compile_state = CompileState::kCompiling;
+  {
+    mutex_lock lock(async_compilation_.async_compilation_mu_);
+    async_compilation_.num_ongoing_compilations++;
+  }
+  // Don't move the above code into the thread function!!!
+
+  // Do not add additional variables to the capture list,
+  // but add any additional ones to the AsyncCompileData struct above
+  // When the ThreadPool for the compilation cache is destroyed, it waits for
+  // compilations to have finished. This means that 'async_compile_data.entry'
+  // and 'this' will be alive for the duration of the compilation.
+  async_compilation_.compiler_threads.get()->Schedule(
+    [this, async_compile_data] {
+      Entry local_entry;
+      VLOG(2) << "Starting asynchronous compilation of cluster "
+              << async_compile_data.function_name << '.';
+      (void)CompileStrict(&local_entry, async_compile_data.options,
+                          async_compile_data.args,
+                          async_compile_data.function_name,
+                          async_compile_data.compile_fn);
+      VLOG(2) << "Finished asynchronous compililation of cluster "
+              << async_compile_data.function_name << '.';
+      {
+        mutex_lock lock(async_compilation_.async_compilation_mu_);
+        async_compilation_.num_ongoing_compilations--;
+      }
+      { // Populate original entry with compilation result.
+        Entry* entry = async_compile_data.entry;
+        mutex_lock entry_lock(entry->mu);
+        entry->compilation_result = local_entry.compilation_result;
+        entry->compile_state = local_entry.compile_state;
+        entry->compilation_status = local_entry.compilation_status;
+        entry->executable = std::move(local_entry.executable);
+      }
+    }
+  );
+  return Status::OK();
+}
+
 Status XlaCompilationCache::CompileImpl(
     const XlaCompiler::Options& options, const NameAttrList& function,
-    absl::Span<const XlaCompiler::Argument> args,
+    const std::vector<XlaCompiler::Argument>& args,
```

Comment:
I don't remember if I asked this earlier, but why did you move to using `std::vector` instead of `absl::Span`?

Reply:
You did ask earlier. args starts out as a vector. it was then converted to a span, but those can't be copied by value.
This needs to be copied by value to be passed to the asynchronous compilation.
If it's a span, it has be turned into a vector first, and then captured by value to the lambda: two copies!!

Reply:
> and then captured by value to the lambda: two copies!!

If you use a `struct` functor then you could `std::move` into the member.  But we've gone around in circles for a while, let's get this checked in and then I'll separately try to refactor if it makes sense.

Reply:
that is dangerous, as that would leave a dangling reference to the caller after the function returns.